Fluescent anomaly and anomaly pads

Time-Green opened the pull request:
Adds the fluescent anomaly. It's basically chemistry anomaly. It drains blood, spawns a predetermined smoke and lubes the floor. 

Adds anomaly pads. They're available on the already existing anomaly tech-node. Use an anomaly on it to suspend it on it. They're pretty much decor and don't do anything, other than being assistant bait and maybe teleporting people if they touch it. 

Adds the chemical anomaly pad. Same as normal anomaly pads, but will drain a fluescent anomalies predetermined smoke chemical at 100 chemicals per process. 

They're both available in all lathes, since they're mainly decor. That and so the bartender can have a rum anomaly in his backroom producing 100 reagents every 2 seconds. 

:cl:
add: A subspace disruption was detected! Fluescent anomalies may now spawn on the station. They might suck your blood out of your body, slip you and constantly create special non-hazardous smoke! 
add: Adds the anomaly pad. You can place an anomaly core on it to have the anomaly be released and suspended on it. Try to keep it powered!
add: Adds the chemical anomaly pad. Every fluescent anomaly spawns with a predetermined chemical, it's the one they constantly emit with smoke. Chemical anomaly pads will generate that chemical in massive amounts.
/:cl:

Why: It's in preperation to central plumbing, where I plan to have the super water synthesizer be a fluescent water anomaly on an anomaly pad. 

I might add a way to imprint a different chemical on a fluescent anomaly in another PR. 

![Screenshot_22](https://user-images.githubusercontent.com/7501474/75189178-f142b300-574d-11ea-975e-3288ea389737.png)

Time-Green committed:
adds fluescent anomaly

Time-Green committed:
Merge remote-tracking branch 'upstream/master' into central-plumbing

Time-Green committed:
partially adds the deep synth

Time-Green committed:
Finishes the anomaly pad and chemical anomaly pad

Krysonism commented:
Add a heavy flow version.

Time-Green committed:
fixes bugs

Time-Green commented:
> Add a heavy flow version.

What would it do? It already produces 100 chems per machine tick

Time-Green committed:
greyscales fluescent anomalie and fixes reagents being inconsistent and some runtimes

Pugie commented:
Reactive lube armor when

ExcessiveUseOfCobblestone commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -345,3 +357,57 @@
+
+	var/smoke_chance = 10
+	var/smoke_range = 5
+	var/smoke_volume = 10
```
autodoc

ExcessiveUseOfCobblestone commented on code/game/machinery/anomaly_pad.dm:
```diff
@@ -0,0 +1,119 @@
+
+	var/volume = 1000
+	var/speed = 100
+	var/reagent_type //alright, it'd be gross to make another anomaly var for the fluescent anom, so just track the reagent_type here
```
autodoc

ExcessiveUseOfCobblestone commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -80,6 +84,14 @@
 	if(I.tool_behaviour == TOOL_ANALYZER)
 		to_chat(user, "<span class='notice'>Analyzing... [src]'s unstable field is fluctuating along frequency [format_frequency(aSignal.frequency)], code [aSignal.code].</span>")
 
+/obj/effect/anomaly/proc/Suspend()
```
doc

ExcessiveUseOfCobblestone commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -80,6 +84,14 @@
+	suspended = TRUE
+	countdown.stop()
+
+/obj/effect/anomaly/proc/Unsuspend()
```
doc

ExcessiveUseOfCobblestone commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -345,3 +357,57 @@
+		smoke.set_up(reagents, smoke_range, get_turf(src), TRUE)
+		smoke.start()
+
+/obj/effect/anomaly/fluid/proc/SetReagent(datum/reagent/R)
```
doc

ExcessiveUseOfCobblestone commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -7,6 +7,8 @@
 	anchored = TRUE
 	light_range = 3
+
+	var/suspended = FALSE //true for no move and no magic lightning fire explosions. used by anomaly pads
```
doc

ExcessiveUseOfCobblestone commented:
mc

Time-Green committed:
Merge remote-tracking branch 'upstream/master' into central-plumbing

Time-Green committed:
improves documentation

Time-Green commented:
Autodocced and more variables

ExcessiveUseOfCobblestone commented:
https://github.com/tgstation/tgstation/blob/master/.github/AUTODOC_GUIDE.md

Time-Green commented:
I know how to autodoc I just forgot to doc the procs

Time-Green committed:
adds autodoc

Time-Green commented:
uh ok I think that was it

Time-Green committed:
unforgets a variable

Time-Green committed:
Doccs types

Time-Green committed:
Adds fluescent anomaly type doc

Rohesie commented on code/game/machinery/anomaly_pad.dm:
```diff
@@ -0,0 +1,125 @@
+	var/base_power = 500
+
+/obj/machinery/anomaly_pad/RefreshParts()
+	..()
```
```suggestion
	. = ..()
```

Time-Green replied:
It doesn't expect a return value

Rohesie replied:
We tend to save the return value nonetheless as a convention. Makes it easier for mass-replacements or for changing the logic down the line, without having to worry if a stray proc forgot to do it.

Time-Green replied:
Alright then, I didn't do I thought it was misleading. This one didn't actually have a sensible parent call, but I'll change the others then

Rohesie commented on code/game/machinery/anomaly_pad.dm:
```diff
@@ -0,0 +1,125 @@
+		playsound(loc, 'sound/machines/click.ogg', 15, TRUE)
+		return
+
+	. = ..()
```
```suggestion
	return ..()
```

Time-Green replied:
I still think this is way sexier, but ok then

Rohesie commented on code/game/machinery/anomaly_pad.dm:
```diff
@@ -0,0 +1,125 @@
+	reagents.add_reagent(reagent_type, speed)
+
+/obj/machinery/anomaly_pad/liquid/CaptureAnomaly(obj/item/assembly/signaler/anomaly/fluid/S)
+	..(S)
```
```suggestion
	. = ..()
```
No need to pass the `S` to the parent, that will already be done by default.

Time-Green replied:
Also doesnt expect a return value, but I'll remove the S

Rohesie commented on code/game/machinery/anomaly_pad.dm:
```diff
@@ -0,0 +1,125 @@
+	use_power = IDLE_POWER_USE
+
+///Check if an anomaly is applicable. Example is the chemical pad only accepting fluescent anomalies
+/obj/machinery/anomaly_pad/proc/IsAnomalyApplicable(obj/item/assembly/signaler/anomaly/S, mob/living/user)
```
```suggestion
/obj/machinery/anomaly_pad/proc/IsAnomalyApplicable(obj/item/I, mob/living/user)
```

Time-Green replied:
Object type is checked before IsAnomalyApplicable is called, and I see no reason to downgrade it. The reason it's still checked later is because that needs to specificly be a fluid anomaly.

Rohesie commented on code/game/machinery/anomaly_pad.dm:
```diff
@@ -0,0 +1,125 @@
+	..(S)
+	reagent_type = S.reagent_type
+
+/obj/machinery/anomaly_pad/liquid/IsAnomalyApplicable(obj/item/assembly/signaler/anomaly/S, mob/living/user)
```
```suggestion
/obj/machinery/anomaly_pad/liquid/IsAnomalyApplicable(obj/item/S, mob/living/user)
```
Keep the incoming input and the type matching here, not to pass incorrect information.

Time-Green replied:
Ditto, incoming is always at least obj/item/assembly/signaler/anomaly

Rohesie commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -347,3 +362,72 @@
+	SetReagent(get_random_reagent_id())
+
+/obj/effect/anomaly/fluid/anomalyEffect()
+	..()
```
```suggestion
	. = ..()
```

Time-Green replied:
It doesn't expect a return value, don't wanna be misleading

Rohesie commented on code/modules/assembly/signaler.dm:
```diff
@@ -211,6 +211,11 @@
+///so we can properly override this in-case we need some special stuff passed along with it, like the reagent type with fluescent anomalies. also pass turf because qdel and
+/obj/item/assembly/signaler/anomaly/proc/ReviveAnomaly(turf/T)
+	qdel(src)
+	return new anomaly_type(T)
```
Have you tested this locally to work properly?

Time-Green replied:
It works properly, but you're right. Running code on qdelled objects feels gross. I'll just use a QDEL_IN with a 0 timer

Time-Green committed:
removes unneeded stuff

Time-Green committed:
Update code/game/machinery/anomaly_pad.dm

Co-Authored-By: Rohesie <rohesie@gmail.com>

Time-Green committed:
changes ..() to . = ..()

(even though theres no return value, but a maintainer told me so dont blame me)

Rohesie commented on code/game/machinery/anomaly_pad.dm:
```diff
@@ -0,0 +1,124 @@
+		ReleaseAnomaly()
+		return FALSE //uuhhh dont know if its ok but im using this to see if the check passed on the subtype
+
+	return TRUE
```
The return here is only relevant if it is `PROCESS_KILL`, in where it stops processing.
Could you make it start processing on `CaptureAnomaly()` and stop on `ReleaseAnomaly()`?

Rohesie commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -80,6 +85,16 @@
+/obj/effect/anomaly/proc/Unsuspend()
+	suspended = FALSE
+	countdown.start()
+
```
Could you make the anomaly stop processing on `Suspend()` and restart on `Unsuspend()`?

Rohesie commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -347,3 +362,73 @@
+/obj/effect/anomaly/fluid/anomalyEffect()
+	. = ..()
+
+	if(lube_chance)
```
```suggestion
	if(prob(lube_chance))
```
I suspect you meant this.

Time-Green replied:
Indeed I did

Time-Green committed:
makes anomalies and anomaly pads stop and start processing when appropriate

Time-Green committed:
Update code/game/objects/effects/anomalies.dm

Co-Authored-By: Rohesie <rohesie@gmail.com>

Time-Green commented:
Thank you for the thorough review

Rohesie commented on code/game/machinery/anomaly_pad.dm:
```diff
@@ -17,6 +17,11 @@
+/obj/machinery/anomaly_pad/Initialize()
+	. = ..()
+
+	STOP_PROCESSING(SSmachines, src)
```
This will conflict with the standard set by https://github.com/tgstation/tgstation/pull/49307 so whichever gets merged first will have to be accounted by the other.

Time-Green replied:
Thanks for the warning, I'll either fix it or warn the other guy if this gets merged first

optimumtact commented:
Not going to merge this personally, dont' like the idea of infinite chems coming from wacky portals.

best of luck

Rohesie commented:
Could you rework, replace or remove the chemical anomaly? It seems not to be in line with the project design.

Time-Green commented:
I made it to start prep on central plumbing, but I'm not continueing it.

I'll close the PR, maybe I'll revive the anomaly pad once I give them a solid purpose. (super sonic station goal)

Rohesie commented:
Up to you. You could get it merged by just removing the chemical anomaly.

Time-Green commented:
Anomaly pads weren’t the focus, just the means. I’ll probably re-PR it later with some depth to it. Maybe make it a mini station-goal for scientists/engineers to do late-game
